daemon/list.go

```diff
@@ -113,44 +115,70 @@ func (daemon *Daemon) Containers(ctx context.Context, config *containertypes.Lis
 		return nil, err
 	}
 
-	// fastpath to only look at a subset of containers if specific name
+	// shortcut to only look at a subset of containers if specific name
 	// or ID matches were provided by the user--otherwise we potentially
 	// end up querying many more containers than intended
 	containerList, err := daemon.filterByNameIDMatches(view, filter)
 	if err != nil {
 		return nil, err
 	}
+	numContainers := len(containerList)
+	results := make(chan *containertypes.Summary, numContainers)
 
+	// Get the info for each container in the list; this can be slow so we
+	// dispatch a set number of worker goroutines to do the jobs. We choose
+	// log2(numContainers) workers to avoid creating too many goroutines
+	// for large number of containers.
+	numWorkers := log2(numContainers)
+	if numWorkers < 1 {
```

Comment:
Just out of curiousity, shall we also limit the maximum workers to the physical cores of the machine ? Since only number of physical cores workers can work concurrently

Reply:
Thanks @coolljt0725 for the comment.

No I don't think we should do that, because I am looking for the tasks to make progress concurrently, but not necessarily run at the exact same time (i.e., simultaneously). Thus we don't need to tie the number of goroutines to the number of CPUs, and instead just let the Golang runtime scheduler take care of scheduling goroutines as needed to make concurrent progress.

Reply:
Perhaps we should benchmark; perhaps we're out-smarting what go itself does. Also looking at image list, where we do this; if it doesn't make an actual difference that definitely makes the code slightly more straightforward

https://github.com/moby/moby/blob/126d4bc65b37353121250a93fea896b3a9364a0d/daemon/containerd/image_list.go#L147-L155